src/setup_payload/AdditionalDataPayloadGenerator.cpp

```diff
@@ -0,0 +1,137 @@
+        sHashInput = nullptr;
+    }
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateAdditionalDataPayload(PacketBufferHandle & bufferHandle,
+                                                                         bool enableRotatingDeviceId)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVWriter writer;
+    TLVWriter innerWriter;
+    char rotatingDeviceIdBuffer[ROTATING_DEVICE_ID_LENGTH * 2 + 1];
+    size_t rotatingDeviceIdBufferSize = 0;
+
+    // Initialize TLVWriter
+    writer.Init(PacketBuffer::New());
+
+    err = writer.OpenContainer(AnonymousTag, kTLVType_Structure, innerWriter);
+    SuccessOrExit(err);
+
+    if (enableRotatingDeviceId)
+    {
+        // Generating Device Rotating Id
+        err = generateRotatingDeviceId(rotatingDeviceIdBuffer, ROTATING_DEVICE_ID_LENGTH * 2 + 1, rotatingDeviceIdBufferSize);
+        SuccessOrExit(err);
+
+        // Adding the rotating device id to the TLV data
+        err = innerWriter.PutString(ContextTag(kRotatingDeviceIdTag), rotatingDeviceIdBuffer);
+        SuccessOrExit(err);
+    }
+
+    err = writer.CloseContainer(innerWriter);
+    SuccessOrExit(err);
+
+    err = writer.Finalize(&bufferHandle);
+    SuccessOrExit(err);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateRotatingDeviceId(char rotatingDeviceIdBuffer[],
+                                                                    size_t rotatingDeviceIdBufferSize,
+                                                                    size_t & rotatingDeviceIdBufferOutputSize)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    char lifetimeCounterBuf[ROTATING_DEVICE_ID_COUNTER_STR_LENGTH];
+
+    uint8_t outputBuffer[ROTATING_DEVICE_ID_LENGTH];
+    uint8_t hashOutputBuffer[kSHA256_Hash_Length];
+    BufferWriter hashInputBufferWriter(sHashInput, ROTATING_DEVICE_ID_COUNTER_STR_LENGTH + mSerialNumberBufferSize);
+    BufferWriter outputBuferWriter(&outputBuffer[0], ArraySize(outputBuffer));
+
+    VerifyOrExit(rotatingDeviceIdBufferSize >= ROTATING_DEVICE_ID_LENGTH * 2 + 1, err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    snprintf(lifetimeCounterBuf, sizeof(lifetimeCounterBuf), "%09u", mLifetimeCounter);
+    hashInputBufferWriter.Put(lifetimeCounterBuf, ROTATING_DEVICE_ID_COUNTER_STR_LENGTH);
+    hashInputBufferWriter.Put(mSerialNumberBuffer, mSerialNumberBufferSize);
+
+    err = Hash_SHA256(hashInputBufferWriter.Buffer(), ROTATING_DEVICE_ID_COUNTER_STR_LENGTH + mSerialNumberBufferSize,
+                      hashOutputBuffer);
+    SuccessOrExit(err);
+
+    // Computing the Rotating Device Id
+    // RDI = Rotation_Counter + SuffixBytes(HMAC_SHA256(Serial_Number + Rotation_Counter), 16)
+    outputBuferWriter.Put(lifetimeCounterBuf);
+    outputBuferWriter.Put(
+        reinterpret_cast<const char *>(&hashOutputBuffer[kSHA256_Hash_Length - ROTATING_DEVICE_ID_HASH_SUFFIX_LENGTH]));
+
+    for (size_t i = 0; i < ROTATING_DEVICE_ID_LENGTH; i++)
```

Comment:
Why are you doing your own hex conversion here? Are there not hex conversion utilities? Why not simply output bytes, and let the caller serialize as hex if needed?

Reply:
i am not aware of any utilities that convert to hex, the rotating device id is hex by specs, so there is no need to let the caller take that responsibility